_posts/2020-07-23-demo-fraud-detection-3.md

```diff
@@ -0,0 +1,655 @@
+us to perform dynamic data partitioning based on the `groupingKeyNames`
+parameter in the rule definition. The subsequent description is focused
+around the `DynamicAlertFunction`, which makes use of the remaining rule
+settings.
+
+<center>
+<img src="{{ site.baseurl }}/img/blog/patterns-blog-3/sample-rule-definition.png" width="700px" alt="Figure 5: Sample Rule Definition"/>
+<br/>
+<i><small>Figure 5: Sample Rule Definition</small></i>
+</center>
+<br/>
+
+
+As described in the previous parts of the blog post
+series, our alerting process function receives events of type
+`Keyed<Transaction, String, Integer>`, where `Transaction` is the main
+"wrapped" event, String is the key (*payer \#x - beneficiary \#y* in
+Figure 1), and `Integer` is the ID of the rule that caused the dispatch of
+this event. This rule was previously [stored]({{ site.baseurl }}/news/2020/03/24/demo-fraud-detection-2.html#broadcast-state-pattern) in the broadcast state and has to be retrieved from that state by the ID. Here is the
+outline of the implementation:
+
+```java
+public class DynamicAlertFunction
+    extends KeyedBroadcastProcessFunction<
+        String, Keyed<Transaction, String, Integer>, Rule, Alert> {
+
+  private transient MapState<Long, Set<Transaction>> windowState;
+
+  @Override
+  public void processElement(
+      Keyed<Transaction, String, Integer> value, ReadOnlyContext ctx, Collector<Alert> out){
+
+    // Add Transaction to state
+    long currentEventTime = value.getWrapped().getEventTime();
+    addToStateValuesSet(windowState, currentEventTime, value.getWrapped());
+
+    // Calculate the aggregate value
+    Rule rule = ctx.getBroadcastState(Descriptors.rulesDescriptor).get(value.getId());
+    Long windowStartTimestampForEvent = rule.getWindowStartTimestampFor(currentEventTime);
+    SimpleAccumulator<BigDecimal> aggregator = RuleHelper.getAggregator(rule);
+
+    for (Long stateEventTime : windowState.keys()) {
+      if (isStateValueInWindow(stateEventTime, windowStartForEvent, currentEventTime)) {
+        aggregateValuesInState(stateEventTime, aggregator, rule);
+      }
+    }
+
+    // Evaluate the rule and trigger an alert if violated
+    BigDecimal aggregateResult = aggregator.getLocalValue();
+    boolean isRuleViolated = rule.apply(aggregateResult);
+    if (isRuleViolated) {
+      long decisionTime = System.currentTimeMillis();
+      out.collect(new Alert<>(rule.getRuleId(),
+                              rule,
+                              value.getKey(),
+                              decisionTime,
+                              value.getWrapped(),
+                              aggregateResult));
+    }
+
+    // Register timers to ensure state cleanup
+    long cleanupTime = (currentEventTime / 1000) * 1000;
+    ctx.timerService().registerEventTimeTimer(cleanupTime);
+  }
+```
+
+<br/>
+Here are the details of the steps:  
+1)  We first add each new event to our window state:
```

Comment:
Can you add these numbers into the code above, e.g. like this? That way, people can better see what it relates to.

For example by marking lines with `// <--- #...` or like this (or maybe you have some other idea):
```
// 1) Add Transaction to state
...
// Calculate the aggregate value
// 2) retrieve broadcased rule
```